compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -56,7 +56,8 @@ pub(crate) fn closure_saved_names_of_captured_variables<'tcx>(
 /// Construct the MIR for a given `DefId`.
 fn mir_build(tcx: TyCtxt<'_>, def: LocalDefId) -> Body<'_> {
     // Ensure unsafeck and abstract const building is ran before we steal the THIR.
-    tcx.ensure_with_value().thir_check_unsafety(def);
+    tcx.ensure_with_value()
```

Comment:
Unrelated to this PR, but why is this `ensure_with_value` :thinking:

Reply:
`ensure` verifies that the dependency is green, without computing anything if not required.
This means that a query that is `ensure`d may need to be computed later if some code needs the result value.

`ensure_with_value` verifies that the result is available, computed or loadable from on-disk cache.

Reply:
Hm.. this means that query side effects can be emitted twice? Once when marking a query green and once when executing the query later to recover the value. I guess we need to make sure that's benign.

Reply:
This is ok for diagnostics, as they are deduplicated. But this is a blocker for replaying any other side-effect. I've been trying to fix this to re-create DefId, but I'm not satisfied by my design.

Reply:
There's an option to turn it off though. There's also extra logic for the parallel compiler to avoid emitting side effects multiple times, which is kind of unnecessary then.

How does DefId relate to this?

Representing side effects as forcible queries would avoid this issue, but does require work to support forcing arbitrary keys.